### DISCUSSION TITLE
Reference lost in pageTransition>onAfterEnter hook

### DISCUSSION THREAD

**Discussion by kevinlaw91** [nuxt/nuxt#19049] [Questions]
Project with global page transition enabled.
I want to access a ref inside `onAfterEnter` hook of a page. However the ref/variables are not defined inside the hook callback.
How can I access the ref and variable?
```html
<template>
   <section>
      <div ref="r">Heading</div>
   </section>
</template>
```

``` js
<script setup>
const r = ref();

const foo = "bar";

onMounted(() => {
  console.log('Mounted!');
  console.log('ref after mount', r.value); // Check console, already mounted
});

definePageMeta({
  pageTransition: {
    onAfterEnter: (el) => {
      console.log('afterEnter!');
      console.log('ref after afterEnter', r.value); // Uncaught ReferenceError: r is not defined

      console.log('foo', foo); // Even this doesn't work
    },
  },
});
</script>
```

**Comment by Baroshem** [comment]
Hey,

Most probably onAfterEnter is happing in a different lifecycle stage than your ref in script setup. In here, you are just passing a function that should be triggered once the page was transitioned so you will not have access to these variables.

**Reply by kevinlaw91** [reply]
Yes I believe so. Is there a workaround to get the refs? This use case is very common and I hope Nuxt can support it.

**Reply by Baroshem** [reply]
Maybe @harlan-zw would be able to help with this

**Reply by harlan-zw** [reply]
It's important to note that `definePageMeta` is a _compiler time_ function. Meaning it gets removed from the Vue SFC at compile time.

> definePageMeta is a compiler macro that you can use to set metadata for your page components located in the pages/ directory (unless [set otherwise](https://nuxt.com/docs/api/configuration/nuxt-config#pages)). This way you can set custom metadata for each static or dynamic route of your Nuxt application.

That means out of the box it can't access anything outside of its scope. There may be limited hoisting support but it won't support refs afaik.

There's documentation available for what it seems like you're trying to do https://nuxt.com/docs/getting-started/transitions#dynamic-transitions

**Reply by Baroshem** [reply]
Thanks @harlan-zw for verifying that out!

@kevinlaw91 does this answer your question? :)
